oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/ModifyMetricUpdater.java

```diff
@@ -0,0 +1,116 @@
+
+import org.apache.jackrabbit.oak.plugins.document.Collection;
+import org.apache.jackrabbit.oak.plugins.document.Document;
+import org.apache.jackrabbit.oak.plugins.document.DocumentStore;
+import org.apache.jackrabbit.oak.plugins.document.DocumentStoreStatsCollector;
+import org.apache.jackrabbit.oak.stats.MeterStats;
+import org.apache.jackrabbit.oak.stats.TimerStats;
+
+import java.util.function.Consumer;
+import java.util.function.ObjIntConsumer;
+import java.util.function.Predicate;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Abstract class to update the metrics for {@link DocumentStoreStatsCollector#doneFindAndModify(long, Collection, String, boolean, boolean, int)} for underlying {@link DocumentStore}
+ *
+ * <p>Concrete implementations provide instances of {@link MeterStats}, {@link TimerStats} based on whether throttling is ongoing or not
+ */
+public abstract class ModifyMetricUpdater {
+
+    private final MeterStats createNodeUpsertMeter;
+    private final TimerStats createNodeUpsertTimer;
+    private final MeterStats updateNodeMeter;
+    private final TimerStats updateNodeTimer;
+    private final MeterStats updateNodeRetryCountMeter;
+    private final MeterStats updateNodeFailureMeter;
+
+    public ModifyMetricUpdater(final MeterStats createNodeUpsertMeter,
+                               final TimerStats createNodeUpsertTimer,
+                               final MeterStats updateNodeMeter,
+                               final TimerStats updateNodeTimer,
+                               final MeterStats updateNodeRetryCountMeter,
+                               final MeterStats updateNodeFailureMeter) {
+        this.createNodeUpsertMeter = createNodeUpsertMeter;
+        this.createNodeUpsertTimer = createNodeUpsertTimer;
+        this.updateNodeMeter = updateNodeMeter;
+        this.updateNodeTimer = updateNodeTimer;
+        this.updateNodeRetryCountMeter = updateNodeRetryCountMeter;
+        this.updateNodeFailureMeter = updateNodeFailureMeter;
+    }
+
+    public void update(final Collection<? extends Document> collection, final int retryCount,
+                       final long timeTakenNanos, final boolean isSuccess, final boolean  newEntry,
+                       final Predicate<Collection<? extends Document>> isNodesCollection,
+                       final StatsConsumer<MeterStats, TimerStats> createStatsConsumer,
+                       final StatsConsumer<MeterStats, TimerStats> updateStatsConsumer,
+                       final ObjIntConsumer<MeterStats> retryNodesConsumer,
+                       final Consumer<MeterStats> failureNodesConsumer) {
+
+        requireNonNull(isNodesCollection);
+        requireNonNull(createStatsConsumer);
+        requireNonNull(updateStatsConsumer);
+        requireNonNull(retryNodesConsumer);
+        requireNonNull(failureNodesConsumer);
+
+        if (isNodesCollection.negate().test(collection)) {return;}
```

Comment:
Can you please reformat and put return on a new line?

Reply:
Sure, will update the code.